### PR TITLE
fix tests for render images from data sources

### LINKED ISSUE #7110: Accepting wrong answer
## Issue Description
- when solving the "Render Images from Data Sources" challenge
- write following code in solution

```
html+="<img src='>"+val.imageLink+"'>";
```

It will not display any image but is still accepted as answer.

### PR DESCRIPTION
**Pull request by nishant-tomer**
#### Pre-Submission Checklist

- [x] Your pull request targets the `staging` branch of FreeCodeCamp.
- [x] Branch starts with either `fix/`, `feature/`, or `translate/` (e.g. `fix/signin-issue`)
- [x] You have only one commit (if not, [squash](https://github.com/FreeCodeCamp/FreeCodeCamp/wiki/git-rebase#squashing-multiple-commits-into-one) them into one commit).
- [x] All new and existing tests pass the command `npm run test-challenges`. Use `git commit --amend` to amend any fixes.
#### Type of Change

- [x] Small bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds new functionality)
- [ ] Breaking change (fix or feature that would change existing functionality)
- [ ] Add new translation (feature adding new translations)
#### Checklist:

- [x] Tested changes locally.
- [x] Closes currently open issue: Closes #7110
#### Description

Earlier this test was just checking that you used val.imageLink in your code and not that you constructed a valid image tag. 
Fix includes implementing a more thorough regex that checks for a validly constructed image tag.

### CODE REVIEW

**Comment by crashuniverse** [comment]
Feedback:  
The regex `html\\s*(\\+=|=\\s*html\\s*\\+\\s*)\\s*(\"|')\\s*<\\s*img\\s+src\\s*=\\s*(?:\"\\s*'|'\\s*\")\\s*\\+\\s*val\\.imageLink\\s*\\+\\s*(?:\"\\s*'|'\\s*\")\\s*>\\s*(?:\"|')\\s*;\\s*` doesn't match the correct code `html+="";`

Can you please find a way to with Save this regex feature somewhere on Internet (it can be regexpal.com) to show why this regex will match a few variations of code.

If I were you I'd explain it in 2 steps: 1 - without unimportant spaces (\\s*) and then 2 - we bring in the army of spaces.

Also, in future I'd like to establish a point somewhere that we should allow \\s and it is really sufficient. Someone who write `<meta title="a book">` is way better than `<meta   title="a book">`

Thanks for your PR. It is important that we make a progress here. @nishant-tomer

**Comment by nishant-tomer** [comment]
@crashuniverse 

Yes, ```regex.test()``` fits the use case perfectly.  will make the changes.

Splitting it into two regex is also a good idea for readability and maintenance.I think we should keep both regex as this challenge is about object property access rather than using += operator. 

and about ```to show why this regex will match a few variations of code.``` and ```<meta title="a book">``` is way better than ```<meta title="a book">```  din't get what are you saying? remove the spaces?

**Comment by crashuniverse** [comment]
Cool, lets make the change for `.test()` and also you can delete line 345 since it is blank line.
Agree w/ you on duplicate regex, lets drop that idea, i like your approach.

You got it right about spaces in `<meta title="a book">`, extra spaces illustrated by me in md actually doesn't show up so both variations I write looked the same. I meant to say, extra spaces don't help much in code. Let's keep `\s*` that you have.

Coming to main point, it was easy for PR reviewer if you include a proof that this regex matches the correct code. When I checked, I found it fails and so is not a correct solution (and I could be wrong if I am setting up things incorrectly). Whatsoever this is a problem for me, and a possible next reviewer, and if we can't QA your work, it doesn't make progress. So think of sharing a proof of rex positive match here. For now I can do this, for example, here: http://www.regexpal.com/?fam=95096

**Comment by nishant-tomer** [comment]
Oh got it. 
About that valid string not matching. you are testing with regexpal probably.

from [MDN docs](https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Global_Objects/String/match)
```regexp
Argument passed to match is a regular expression object. If a non-RegExp object obj is passed, it is implicitly converted to a RegExp by using new RegExp(obj).
```
While parsing the argument to ```String.match(obj)```, ```RegExp(obj)``` function treats it as a literal string and removes the escape character ```\``` and if we use just one backslash, in a case like ```\+```, our escaped character becomes meta character ```+``` . In regexpal or regex101 or such, you need to remove extra backslash from everywhere for it to work.

**Comment by crashuniverse** [comment]
Correct, I tried dev console as well. You can maybe help me testing this?

Two new things:
1. You may have to update your branch with things we improve on.
2. You should also rebase your branch w/ `freecodecamp/staging` to avoid any other conflicts.

**Comment by nishant-tomer** [comment]
Here,
Its working.
http://www.regexpal.com/?fam=95097

**Comment by nishant-tomer** [comment]
Oh, Just tested.
Even with ```regexLiteral.test```, we have to escape doubly. In this case also, Engine will implicitly call ```new Regex(regexLiteral)```

**Comment by crashuniverse** [comment]
OK. We can keep `.match()`. Lets take up `.test()` in a different PR to keep it consistent with everything else on the page.

Please do delete a blank line. I can see the amend.

File: seed/challenges/01-front-end-development-certification/json-apis-and-ajax.json [hunk 1]
```diff
@@ -340,7 +340,9 @@
         "</div>"
       ],
       "tests": [
-        "assert(code.match(/val.imageLink/gi), 'message: You should have used the <code>imageLink</code> property to display the images.');"
+        "assert(code.match(/val.imageLink/g), 'message: You should use the <code>imageLink</code> property to display the images.');",
+        "assert(code.match(/html\\s*(\\+=|=\\s*html\\s*\\+\\s*)\\s*(\"|')\\s*<\\s*img\\s+src\\s*=\\s*(?:\"\\s*'|'\\s*\")\\s*\\+\\s*val\\.imageLink\\s*\\+\\s*(?:\"\\s*'|'\\s*\")\\s*>\\s*(?:\"|')\\s*;\\s*/gi), 'message: You should use a valid HTML img tag.');"
+
```
Comment (crashuniverse) [inline seed/challenges/01-front-end-development-certification/json-apis-and-ajax.json]: This needs to go away, sorry! 😓

Comment (nishant-tomer) [inline seed/challenges/01-front-end-development-certification/json-apis-and-ajax.json]: Yeah, My bad. Silly mistake. I deleted it but forgot to add updated file to git staging before commiting.

**Comment by crashuniverse** [comment]
Illustration of Regex work with removing escape characters `\` for RegexPal: http://www.regexpal.com/?fam=95097

This shows that regex works.

**Comment by crashuniverse** [comment]
LGTM

File: seed/challenges/01-front-end-development-certification/json-apis-and-ajax.json [hunk 2]
```diff
@@ -340,7 +340,8 @@
         "</div>"
       ],
       "tests": [
-        "assert(code.match(/val.imageLink/gi), 'message: You should have used the <code>imageLink</code> property to display the images.');"
+        "assert(code.match(/val.imageLink/g), 'message: You should use the <code>imageLink</code> property to display the images.');",
+        "assert(code.match(/html\\s*(\\+=|=\\s*html\\s*\\+\\s*)\\s*(\"|')\\s*<\\s*img\\s+src\\s*=\\s*(?:\"\\s*'|'\\s*\")\\s*\\+\\s*val\\.imageLink\\s*\\+\\s*(?:\"\\s*'|'\\s*\")\\s*>\\s*(?:\"|')\\s*;\\s*/gi), 'message: You should use a valid HTML img tag.');"
```
Comment (Bouncey) [inline seed/challenges/01-front-end-development-certification/json-apis-and-ajax.json]: In the `message`, can you wrap `img` in `<code></code>` tags please.

Comment (Bouncey) [inline seed/challenges/01-front-end-development-certification/json-apis-and-ajax.json]: `html = "";` passes the challenge, but the challenge asks for us to loop through the images.

This only displays the last image.

Comment (nishant-tomer) [inline seed/challenges/01-front-end-development-certification/json-apis-and-ajax.json]: > this only displays the last image.
No. It display all the images after you click `Get Message` Button.

Is that what you are saying? or i am missing something?

Comment (Bouncey) [inline seed/challenges/01-front-end-development-certification/json-apis-and-ajax.json]: Sorry, I should have been clearer, your regex passes `html +=` and `html =`.

If `html =` is used, then only one image is displayed after pressing `Get Message`.

The challenge is to loop through the object and display the images. This then leads on to the next challenge where we filter out some images.

**Comment by raisedadead** [comment]
@nishant-tomer any updates on this?

**Comment by BerkeleyTrue** [comment]
closing as stale.

Feel free to re-open when ready.

Happy Coding
